README.md
# async-local-storage

[![Build Status](https://travis-ci.org/vicanso/async-local-storage.svg?branch=master)](https://travis-ci.org/vicanso/async-local-storage)
[![Coverage Status](https://img.shields.io/coveralls/vicanso/async-local-storage/master.svg?style=flat)](https://coveralls.io/r/vicanso/async-local-storage?branch=master)
[![npm](http://img.shields.io/npm/v/async-local-storage.svg?style=flat-square)](https://www.npmjs.org/package/async-local-storage)
[![Github Releases](https://img.shields.io/npm/dm/async-local-storage.svg?style=flat-square)](https://github.com/vicanso/async-local-storage)

I want something like `thread-local` storage in threaded programming and `async_hooks` is usable in `node.js 8.0`, so there is an easy way to use `thread-local`.

## API

```js
const als = require('async-local-storage');
als.enable();
setTimeout(() => {
  als.scope();
  const id = randomBytes(8);
  als.set('id', id);
  delay().then(() => {
    assert.equal(als.get('id'), id);
    return readfilePromise(__filename);
  }).then(() => {
    assert.equal(als.get('id'), id);
    return superagent.get('http://www.baidu.com/');
  }).then(() => {
    assert.equal(als.get('id'), id);
  });
}, 100);
```

### enable

enable the async hooks

```js
const als = require('async-local-storage');
als.enable();
```

### disable

disable the async hooks

```js
const als = require('async-local-storage');
als.enable();
setTimeout(() => {
  als.disable();
}, 100);
```

### size

get the size of storage

```js
const als = require('async-local-storage');
als.enable();
setTimeout(() => {
  console.info(als.size());
}, 100);
```

### scope

change the scope of call chain, it will be the call chain top (remove the parent of itself)

```js
const als = require('async-local-storage');
const Koa = require('koa');
const assert = require('assert');

const app = new Koa();
app.use(async (ctx, next) => {
  const id = ctx.get('X-Request-Id');
  als.scope();
  als.set('id', id);
  await next();
});

app.use(async (ctx, next) => {
  const id = ctx.get('X-Request-Id');
  assert.equal(als.get('id'), id);
  await next();
});

app.use((ctx) => {
  ctx.body = 'OK';
});
```


### set

set the value by key for the current id

- `key` the key
- `value` the value
- `linkedTop` set the value linked to top

```js
als.enable()
setTimeout(() => {
  als.scope();
  const id = randomBytes();
  setTimeout(() => {
    als.set('id', id, true);
  }, 1);
  setTimeout(() => {
    assert.equal(als.get('id'), id);
  }, 10);
}, 10);
```

### get

get the value by key, if will find from parent, self --> parent --> parent, until the value is not undefined

- `key` the key

```js
als.enable();
setTimeout(() => {
  als.scope();
  const id = randomBytes();
  setTimeout(() => {
    als.set('id', id, true);
  }, 1);
  setTimeout(() => {
    assert.equal(als.get('id'), id);
  }, 10);
}, 10);
```

### enableLinkedTop

enable linked top for default (default is disabled)

```js
als.enable();
als.enableLinkedTop();
setTimeout(() => {
  als.scope();
  setTimeout(() => {
    // the same as als.set('id', 'a', true)
    als.set('id', 'a');
  }, 10);
}, 10);
```

### disableLinkedTop

disable linked top for default

```js
als.enable();
als.enableLinkedTop();
setTimeout(() => {
  als.disableLinkedTop();
  als.scope();
  setTimeout(() => {
    // the same as als.set('id', 'a', false)
    als.set('id', 'a');
  }, 10);
}, 10);
```


### currentId

get the current id

```js
const assert = require('assert');
als.enable();
setTimeout(() => {
  console.info(als.currentId());
}, 10);
```

### use

get the use time of id

- `id` The tigger id, default is `als.currentId()`

```js
als.enable()
setTimeout(() => {
  const id = als.currentId();
  console.info(als.use(id));
}, 10);
```

### enableCreateTime

enable create time of data, default is enabled.

```js
als.enableCreateTime();
```

### disableCreateTime

disable create time of data, it can save memory.

```js
als.disableCreateTime();
```


## TODO
* [ ] 拆分数据和父子对应关系


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [2.3.4](https://github.com/vicanso/async-local-storage/compare/v2.3.3...v2.3.4) (2020-04-26)

### [2.3.3](https://github.com/vicanso/async-local-storage/compare/v2.3.2...v2.3.3) (2020-04-26)

### [2.3.2](https://github.com/vicanso/async-local-storage/compare/v2.3.1...v2.3.2) (2020-04-24)


### Bug Fixes

* delay remove resolvePromise ([ca1ac31](https://github.com/vicanso/async-local-storage/commit/ca1ac3153750ff659758754605cb7490548ebb89))


als.test.js
const path = require('path')
const fs = require('fs')
const util = require('util')
const crypto = require('crypto')
const dns = require('dns')
const http = require('http')
const request = require('supertest')
const assert = require('power-assert')

const als = require('./als')
const { server, redisClient } = require('./support/server')

const topList = []

const delay = ms => new Promise(resolve => setTimeout(resolve, ms))
const randomBytes = () => crypto.randomBytes(8).toString('hex')
const writeFile = util.promisify(fs.writeFile)
const readFile = util.promisify(fs.readFile)
const lookup = util.promisify(dns.lookup)
const lookupService = util.promisify(dns.lookupService)
const file = path.join(__dirname, './tmp')
fs.writeFileSync(file, '')

als.enable()

/* eslint-env jest */
afterAll(() => {
  als.disable()
  redisClient.disconnect()
  server.close()
  const ids = []
  let err = null
  topList.forEach(item => {
    if (ids.indexOf(item.id) !== -1) {
      err = new Error('the id should be unique')
    } else {
      ids.push(item.id)
    }
  })
  if (err) {
    console.error(err)
    throw err
  }
})

describe('fs', () => {
  test('fs watch', async () => {
    als.scope()
    const id = randomBytes()
    als.set('id', id)
    const watcher = fs.watch(file, () => {
      expect(als.get('id')).toBe(id)
      watcher.close()
    })
    await writeFile(file, 'a')
    topList.push(als.top())
  })

  test('fs write/read promise', () => {
    als.scope()
    const id = randomBytes()
    als.set('id', id)
    return writeFile(file, 'a')
      .then(() => {
        expect(als.get('id')).toBe(id)
        return new Promise((resolve, reject) => {
          fs.readFile(file, err => {
            if (err) {
              reject(err)
              return
            }
            expect(als.get('id')).toBe(id)
            resolve()
          })
        })
      })
      .then(() => {
        topList.push(als.top())
      })
  })

  test('fs write/read async/await', async () => {
    als.scope()
    const id = randomBytes()
    als.set('id', id)
    await writeFile(file, 'a')
    expect(als.get('id')).toBe(id)
    await readFile(file)
    expect(als.get('id')).toBe(id)
    topList.push(als.top())
  })
})

describe('JSSTREAM', () => {
  test('js read stream', () => new Promise((resolve, reject) => {
    als.scope()
    const id = randomBytes()
    als.set('id', id)
    const stream = fs.createReadStream(file)
    stream.on('data', () => {
      setTimeout(() => {
        expect(als.get('id')).toBe(id)
        stream.close()
      })
    })
    stream.on('error', reject)
    stream.on('close', () => {
      setTimeout(() => {
        topList.push(als.top())
        expect(als.get('id')).toBe(id)
        resolve()
      })
    })
    fs.writeFileSync(file, 'ab')
  }))

  test('js write stream', () => new Promise((resolve, reject) => {
    als.scope()
    const id = randomBytes()
    als.set('id', id)
    const stream = fs.createWriteStream(file)
    stream.on('error', reject)
    stream.on('close', () => {
      setTimeout(() => {
        topList.push(als.top())
        expect(als.get('id')).toBe(id)
        resolve()
      })
    })
    stream.write('ab')
    stream.end()
  }))
})

describe('dns', () => {
  test('get addr info promise', async () => {
    als.scope()
    const id = randomBytes()
    const asyncId = als.currentId()
    als.set('id', id)
    await lookup('www.baidu.com')
      .then(() => {
        assert(als.get('id') === id)
        assert(als.currentId() !== asyncId)
      })
      .then(() => {
        topList.push(als.top())
        assert(als.get('id') === id)
        assert(als.currentId() !== asyncId)
      })
    assert(als.currentId() !== asyncId)
    assert(als.get('id') === id)
  })

  test('get addr info async/await', async () => {
    als.scope()
    const id = randomBytes()
    als.set('id', id)
    await lookup('www.baidu.com')
    expect(als.get('id')).toBe(id)
    topList.push(als.top())
  })

  test('get name info promise', () => {
    als.scope()
    const id = randomBytes()
    als.set('id', id)
    return lookupService('127.0.0.1', 80)
      .then(() => {
        expect(als.get('id')).toBe(id)
      })
      .then(() => {
        topList.push(als.top())
      })
  })

  test('get name info async/await', async () => {
    als.scope()
    const id = randomBytes()
    als.set('id', id)
    await lookupService('127.0.0.1', 80)
    expect(als.get('id')).toBe(id)
    topList.push(als.top())
  })
})

describe('http', () => {
  test('http get', () => {
    als.scope()
    const id = randomBytes()
    als.set('id', id)
    return new Promise((resolve, reject) => {
      expect(als.get('id')).toBe(id)
      http
        .get('http://www.baidu.com/', res => {
          expect(als.get('id')).toBe(id)
          res.on('data', () => {
            expect(als.get('id')).toBe(id)
          })
          res.on('end', () => {
            expect(als.get('id')).toBe(id)
            resolve()
          })
        })
        .on('error', reject)
    }).then(() => {
      expect(als.get('id')).toBe(id)
      topList.push(als.top())
    })
  })
})

describe('linked top', () => {
  test('normal linked top', () => {
    als.scope()
    const id = randomBytes()
    const user = randomBytes()
    als.set('id', id)
    return delay(10)
      .then(() => {
        const fn1 = delay(10).then(() => {
          const current = als.getCurrentData()
          expect(current).toBeUndefined()
        })
        const fn2 = new Promise(resolve => {
          als.set('user', user)
          resolve()
        })
        return Promise.all([fn1, fn2])
      })
      .then(() => {
        expect(als.get('id')).toBe(id)
        expect(als.get('user')).toBe(user)
      })
  })

  test('set value to self scope', () => {
    als.scope()
    const id = randomBytes()
    const user = randomBytes()
    als.set('id', id)
    return delay(10)
      .then(() => {
        expect(als.get('id')).toBe(id)
        als.set('id', 'a')
        als.set('user', user)
      })
      .then(() => {
        expect(als.get('id')).toBe('a')
        expect(als.get('user')).toBe(user)
      })
  })

  test('enableLinkedTop', () => {
    als.scope()
    const id = randomBytes()
    const user = randomBytes()
    als.enableLinkedTop()
    als.set('id', id)
    const current = als.getCurrentData()
    return delay(10)
      .then(() => {
        als.set('user', user)
      })
      .then(() => {
        expect(current.id).toBe(id)
        expect(current.user).toBe(user)
        als.disableLinkedTop()
      })
  })
})

describe('get use', () => {
  beforeAll(() => {
    als.enableCreateTime()
  })
  test('use', async () => {
    als.scope()
    let id = als.currentId()
    return delay(10).then(() => {
      const use = als.use(id)
      expect(use).toBeGreaterThanOrEqual(9 * 1000 * 1000)
      id = als.currentId()
      als.disableCreateTime()
      return delay(10)
    }).then(() => {
      const use = als.use(id)
      expect(use).toBe(-1)
    })
  })
})

describe('size', () => {
  test('get size', () => {
    expect(als.size()).toBeGreaterThan(1)
  })
})

describe('koa', () => {
  const check = url => {
    const fns = [1, 2, 3, 4, 5].map(() => {
      const id = randomBytes(8)
      return request(server)
        .get(url)
        .set('X-Request-Id', id)
        .expect(200)
        .then(response => {
          expect(response.text).toBe(id)
        })
    })
    return Promise.all(fns)
  }
  test('get request id', () => check('/'))

  test('get request id (fs)', () => check('/?fs=true'))

  test('get request id(delay)', () => check('/?delay=true'))

  test('get request id(http)', () => check('/?http=true'))

  test('get request id(session)', () => check('/?session=true'))

  test('get request id(all)', () => check('/?fs=true&delay=true&http=true&session=true'))
})

describe('getFromParent', () => {
  test('top', () => {
    als.scope()
    let id = als.getFromParent('id')
    expect(id).toBe(undefined)
    als.set('id', 1)
    id = als.getFromParent('id')
    expect(id).toBe(undefined)
  })
  describe('1 level', () => {
    test('single', () => {
      als.scope()
      als.set('id', 1)
      return delay(10).then(() => {
        expect(als.getFromParent('id')).toBe(1)
        als.set('id', 2)
        expect(als.getFromParent('id')).toBe(1)
        expect(als.get('id')).toBe(2)
      })
    })
    test('multiple', () => {
      als.scope()
      als.set('id', 1)
      return Promise.all([
        delay(10).then(() => {
          expect(als.getFromParent('id')).toBe(1)
          als.set('id', 2)
          expect(als.getFromParent('id')).toBe(1)
          expect(als.get('id')).toBe(2)
        }),
        delay(10).then(() => {
          expect(als.getFromParent('id')).toBe(1)
          als.set('id', 3)
          expect(als.getFromParent('id')).toBe(1)
          expect(als.get('id')).toBe(3)
        })
      ]).then(() => {
        expect(als.get('id')).toBe(1)
      })
    })
  })
  test('2 level', () => {
    als.scope()
    als.set('id', 1)
    return delay(10).then(() => {
      als.set('key2', 1)
      return delay(10).then(() => {
        expect(als.get('key2')).toBe(1)
        expect(als.getFromParent('id')).toBe(1)
        als.set('id', 2)
        expect(als.getFromParent('id')).toBe(1)
        expect(als.get('id')).toBe(2)
      })
    })
  })
})

describe('get all data', () => {
  expect(als.getAllData()).toBeDefined()
})

describe('simple', () => {
  test('top Data', async () => {
    als.scope()
    als.set('name', 'main')
    assert(als.get('name') === 'main')
    await Promise.all([
      delay(200).then(() => {
        assert(als.get('name') === 'main')
        als.set('name', 'sleep1')
      }).then(() => {
        assert(als.get('name') === 'sleep1')
      }),
      delay(500).then(() => {
        assert(als.get('name') === 'main')
      })
    ])
  })
})

describe('loop', () => {
  test('loopTestWhileWithOutScope', async () => {
    let count = 0
    /* eslint-disable no-await-in-loop */
    while (count < 300) {
      count += 1
      expect(als.size() > count)
      await delay(1)
      expect(als.size() > count)
    }
  })
  test('loopWithScope', async () => {
    als.scope()
    als.setRemoveDelay(200)
    let count = 0
    /* eslint-disable no-await-in-loop */
    while (count < 300) {
      expect(als.size() < 20)
      count += 1
      await delay(1)
      als.scope()
      als.set('test', count)
      await delay(1)
      expect(als.get('test')).toBe(count)
    }
  })
})


als.js
const asyncHooks = require('async_hooks')
const nano = require('nano-seconds')
const util = require('util')
const fs = require('fs')
const assert = require('assert')

const enabledDebug = process.env.DEBUG === 'als'
const toRemove = new Map()
const relationMap = new Map()
const dataMap = new Map()

const BLACKLIST_TYPE = [
  'GETADDRINFOREQWRAP',
  'GETNAMEINFOREQWRAP',
  'TCPCONNECTWRAP',
  'WRITEWRAP',
  'TTYWRAP',
  'SIGNALWRAP'
]

let timer
let defaultLinkedTop = false
let enabledCreatedAt = false
let removeDelay = 60000

assert(asyncHooks.executionAsyncId)

const debug = (...args) => {
  if (!enabledDebug) {
    return
  }
  // use a function like this one when debugging inside an AsyncHooks callback
  fs.writeSync(1, `${util.format(...args)}\n`)
}

const getTopData = id => {
  const data = dataMap.get(id)
  if (data && data.isTop) {
    return data
  }
  const parentId = relationMap.get(id)
  if (!parentId) {
    return data
  }
  return getTopData(parentId)
}

const getData = (id, key) => {
  const data = dataMap.get(id) || {}
  const value = data[key]
  if (value !== undefined) {
    return value
  }
  if (!data.isTop) {
    const parentId = relationMap.get(id)
    if (parentId) {
      return getData(parentId, key)
    }
  }
  return undefined
}


const delayRemove = (id, delay = removeDelay) => {
  if (!relationMap.has(id)) {
    return
  }
  debug('delayRemove %d;%d', id, delay)
  toRemove.set(id, { id, removeAt: Date.now() + delay })
}

// https://nodejs.org/dist/latest-v12.x/docs/api/async_hooks.html#async_hooks_init_asyncid_type_triggerasyncid_resource
const hooks = asyncHooks.createHook({
  init: function init (id, type, triggerId) {
    debug('%d(%s) init by %d', id, type, triggerId)
    if (BLACKLIST_TYPE.includes(type)) {
      return
    }
    relationMap.set(id, triggerId)
    if (enabledCreatedAt) {
      dataMap.set(id, { created: nano.now() })
    }
  },
  promiseResolve: delayRemove,
  after: delayRemove,
  destroy: id => delayRemove(id, 5000)
})

/**
 * Get the current id
 * @returns {int} currentId
 */
const getCurrentId = () => asyncHooks.executionAsyncId()

exports.currentId = getCurrentId

/**
 * Enable the async hook
 * @returns {AsyncHook} A reference to asyncHook.
 */
exports.enable = () => {
  timer = setInterval(() => {
    debug('toRemoveSize %d;', toRemove.size)
    const now = Date.now()
    toRemove.forEach(item => {
      const { id, removeAt } = item
      if (removeAt < now) {
        toRemove.delete(id)
        relationMap.delete(id)
        dataMap.delete(id)
        debug('remove %d;', id)
      }
    })
    debug('toRemoveSize %d;', toRemove.size)
  }, 20000)
  return hooks.enable()
}

/**
 * Disable the async hook
 * @returns {AsyncHook} A reference to asyncHook.
 */
exports.disable = () => {
  relationMap.clear()
  dataMap.clear()
  toRemove.clear()
  clearInterval(timer)
  return hooks.disable()
}

/**
 * Get the size of map
 * @returns {int} size
 */
exports.size = () => dataMap.size

/**
 * Enable linked top
 * @returns {void}
 */
exports.enableLinkedTop = () => {
  defaultLinkedTop = true
}

/**
 * Disable linked top
 * @returns {void}
 */
exports.disableLinkedTop = () => {
  defaultLinkedTop = false
}

/**
 * Set the key/value for this score
 * @param {string} key The key of value
 * @param {any} value The value
 * @param {boolean} [linkedTop] The value linked to top
 * @returns {boolean} if success, will return true, otherwise false
 */
exports.set = function setValue (key, value, linkedTop = defaultLinkedTop) {
  /* istanbul ignore if */
  if (key === 'created') {
    throw new Error("can't set created")
  }
  const id = getCurrentId()
  debug('set %s:%j to %d', key, value, id)
  if (linkedTop) {
    const data = getTopData(id)
    if (data) {
      data[key] = value
    }
  }
  const currentData = dataMap.get(id)
  if (!currentData) {
    dataMap.set(id, { [key]: value })
  } else {
    currentData[key] = value
  }
}

/**
 * Get the value by key
 * @param {string} key The key of value
 * @returns {any} value
 */
exports.get = key => getData(getCurrentId(), key)

/**
 * 获取当前current data
 * @returns {object} current data
 */
exports.getCurrentData = () => dataMap.get(getCurrentId())

/**
 * Get the value by key from parent
 * @param {string} key The key of value
 * @returns {any} value
 */
exports.getFromParent = key => getData(
  relationMap.get(getCurrentId()),
  key
)

/**
 * Remove the data of the current id
 * @returns {void}
 */
exports.remove = function removeValue () {
  const id = getCurrentId()
  if (id) {
    dataMap.delete(id)
    relationMap.delete(id)
    toRemove.delete(id)
  }
}

/**
 * Get the use the of id
 * @param {number} id The trigger id, is optional, default is `als.currentId()`
 * @returns {number} The use time(ns) of the current id
 */
exports.use = function getUse (id) {
  const data = dataMap.get(id || getCurrentId())
  /* istanbul ignore if */
  if (!data || !enabledCreatedAt) {
    return -1
  }
  return nano.difference(data.created)
}

/**
 * Get the top value
 * @returns {object} topData
 */
exports.top = function top () {
  return getTopData(getCurrentId())
}

/**
 * Set the scope (it will change the top)
 * @returns {void}
 */
exports.scope = function scope () {
  const id = getCurrentId()
  const currentData = dataMap.get(id)
  if (currentData) {
    currentData.isTop = true
  } else {
    dataMap.set(id, { isTop: true })
  }
}

/**
 * Get all data of async locatl storage, please don't modify the data
 * @returns {map} allData
 */
exports.getAllData = function getAllData () {
  return dataMap
}

/**
 * Enable the create time of data
 * @returns {void}
 */
exports.enableCreateTime = function enableCreateTime () {
  enabledCreatedAt = true
}

/**
 * Disable the create time of data
 * @returns {void}
 */
exports.disableCreateTime = function disableCreateTime () {
  enabledCreatedAt = false
}

exports.setRemoveDelay = delay => {
  removeDelay = delay
}

exports.parentId = () => relationMap.get(getCurrentId())

exports.sizes = () => ({
  toRemove: toRemove.size,
  relationMap: relationMap.size,
  dataMap: dataMap.size
})


examples/loop.js
const als = require('..');
const fs = require('fs');
const assert = require('assert')
const got = require('got');
const debug = require('debug')('als')
const util = require('util');
const dns = require('dns');

const lookup = util.promisify(dns.lookup);
const lookupService = util.promisify(dns.lookupService);

als.enable();
als.setRemoveDelay(5000)

const scope = async () => {
  als.scope();
  als.set('id', Math.random())
  await sleep(300)
}

const sleep = timeout => new Promise(resolve => setTimeout(() => {
  debug('sleep', als.currentId())
  als.set('id', Math.random())
  resolve();
}, timeout))

const request = url => {
  als.set('id', Math.random())
  return got(url, {})
}

let count = 0

const loop = () => {
  setTimeout(async () => {
    while (true) {
      await scope()
      const id = Math.random()
      als.set('id', id)
      debug('while', als.currentId())
      count += 1
      await sleep(400)
      debug('whileDone', als.currentId())
      debug('http>>>>', als.currentId())
      await got('https://www.baidu.com/').then(() => {
        setTimeout(() => {
          debug('http', als.currentId())
        })
      })
      debug('http<<<<', als.currentId())
      await sleep(400)
      debug('lookup>>>>', als.currentId())
      await lookup('www.baidu.com').then(() => {
        setTimeout(() => {
          debug('lookup', als.currentId())
        })
      })
      debug('lookup<<<<', als.currentId())
      await sleep(500)
      debug('lookupService>>>>', als.currentId())
      await lookupService('127.0.0.1', 80).then(() => {
        setTimeout(() => {
          debug('lookupService', als.currentId())
        })
      })
      debug('lookupService<<<<', als.currentId())
      await sleep(500)
      debug('log', als.currentId())
    }
  })
}

setImmediate(() => {
  loop()
  loop()
  loop()
  loop()
})

setInterval(() => {
  debug('===========', count, als.getAllData().keys(), als.size())
  const { toRemove, relationMap, dataMap }= als.sizes()
  console.log('===========', count, als.sizes())
  assert(relationMap - toRemove < 50)
  assert(dataMap < 500)
}, 2000)

/*
setTimeout(() => {
  als.disable()
}, 6000)
*/
